OpenSim/Common/DataTable.h

```
@@ -0,0 +1,274 @@
+
+/** @file
+* This file defines the  DataTable class, which is used by OpenSim to provide
+* an in memory container for data access and manipulation.
+*/
+#include "SimTKmath.h"
+#include "XMLDocument.h"
+#include "FileAdapter.h"
+#include "Array.h"
+#include <memory>
+#include <typeindex>
+
+
+namespace OpenSim {
+
+//=============================================================================
+/**
+ * AbstractDataTable defines a simple interface for in memory access to
+ * numerical data as a table. A DataTable is independent of the data source 
+ * used to populate it. A concrete DataTable provides fast access to data
+ * elements by row and/or column indices as well as columns by label.
+ *
+ * @author Ajay Seth
+ */
+class AbstractDataTable {
+public:
+    virtual ~AbstractDataTable() = default;
+
+    virtual AbstractDataTable* clone() const = 0;
+
+    /** Read the meta data associated with this DataTable as an XMLDocument */
+    virtual const XMLDocument& getMetaData() const = 0;
+    /** Update the meta data associated with this DataTable as an XMLDocument */
+    virtual XMLDocument& updMetaData() = 0;
+
+    /** Get the number of rows in the DataTable */
+    virtual size_t  getNumRows() const = 0;
+    /** Get the number of columns in the DataTable */
+    virtual size_t  getNumCols() const = 0;
+
+    /** Determine if the DataTable has a column associated with a 
+        column label (string) */
+    virtual bool hasColumn(const std::string& label) const = 0;
+    /** For a known column label return its column index. */
+    virtual size_t getColumnIndex(const std::string& label) const = 0;
+
+    /** Read-only column labels associated with the columns of the DataTable */
+    virtual const Array<std::string>& getColumnLabels() const = 0;
+    /** Update column labels associated with the columns of the DataTable */
+    virtual Array<std::string>& updColumnLabels() = 0;
```

I don't think so. We can throw an `InvalidNumCols` exception and add details to the documentation.